aiida_quantumespresso/workflows/pw/relax.py

```diff
@@ -11,6 +11,17 @@
 PwBaseWorkChain = WorkflowFactory('quantumespresso.pw.base')
 
 
+def validate_final_scf(value, _):
```

Comment:
I haven't seen `_` used in this context. What's its purpose here?

Re the use of `value` as the argument: I've seen both `value` used as an argument for a validator:

https://github.com/aiidateam/aiida-quantumespresso/blob/6924ce13b6d02caa8d0ca5e2dc238a1d72ae650e/aiida_quantumespresso/calculations/pp.py#L13-L15

As well as the [name of the input](https://github.com/aiidateam/aiida-quantumespresso-hp/blob/50669e934c76bd99078bb4bacc1056212868ae35/aiida_quantumespresso_hp/calculations/hp.py#L17). (Shame you can't make code snippets from other repo's 😭 ).

What would be the better standard? Using `value` creates a more consistent function definition, but I think using the name of the input is more readable.

Reply:
The `_` character is commonly used as a placeholder for an argument that is not used. It is also accepted by `pylint` so we don;t have to disable the `unused-argument` warning.

Regarding the naming: I think using the general `value` is better, because a validator can theoretically be used for more than one port that have different names. It is also more robust w.r.t. refactoring.

Reply:
> The `_` character is commonly used as a placeholder for an argument that is not used. It is also accepted by `pylint` so we don;t have to disable the `unused-argument` warning.

Thanks for the clarification! So to hash this out a little further: the signature of a validator should adhere to the following format:

```Python
validator(value, port)
```

In case it doesn't, `plumpy` will raise [this](https://github.com/aiidateam/plumpy/blob/87337f649fe2b03e7007301b36e2633b6918fc4b/plumpy/ports.py#L17) deprecation warning. The explanation (found in [this commit](https://github.com/aiidateam/plumpy/commit/2b1a268f8d4fbcbdc40a319fee0e2cae91278382)) is that the port argument can be used in the context of a `PortNameSpace` to see what inputs are actually present. I suppose the reason why the old signature (`validator(value)`) is deprecated is because we want to have one consistent signature for *all* validators?

One comment is that the [documentation on validation](https://aiida.readthedocs.io/projects/aiida-core/en/latest/topics/processes/usage.html#validation-and-defaults) is still outdated with regard to this change in signature. The example used there only has one argument, and doesn't use `value`: 

```Python
def is_number_positive(number):
    if number < 0:
        return 'The number has to be greater or equal to zero'
```
I'd be happy to take a crack at fixing this!

> Regarding the naming: I think using the general `value` is better, because a validator can theoretically be used for more than one port that have different names. It is also more robust w.r.t. refactoring.

👍

Reply:
The reason for deprecating the old signature and requiring validators to define two arguments, is that if you pass two arguments to a validator that only accepts one, you will raise an exception. Of course we can inspect the signature to see what we should pass (as we are doing now to determine whether we should emit a deprecation warning) it is better to just have all validators use the same signature.

Re the out of date documentation: would be great indeed if you could fire off a PR to fix it. Try and include also an explanation on why the second argument is passed and how it should be used. The commit that you referenced, that introduced the new signature, should give you a good hint when it it useful to look at the port.